cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaRoller.java

```diff
@@ -416,22 +447,55 @@ String podName(Integer podId) {
      * @return A future which completes the the node id of the controller of the cluster,
      * or -1 if there is not currently a controller.
      */
-    int controller(int podId, Admin ac, long timeout, TimeUnit unit) throws ForceableProblem, InterruptedException {
+    int controller(int podId, Admin ac, long timeout, TimeUnit unit, RestartContext restartContext) throws ForceableProblem, InterruptedException {
         Node controllerNode = null;
         try {
             DescribeClusterResult describeClusterResult = ac.describeCluster();
             KafkaFuture<Node> controller = describeClusterResult.controller();
             controllerNode = controller.get(timeout, unit);
-        } catch (ExecutionException e) {
-            throw new ForceableProblem("Error while trying to determine the cluster controller from pod " + podName(podId), e.getCause());
-        } catch (TimeoutException e) {
-            throw new ForceableProblem("Error while trying to determine the cluster controller from pod " + podName(podId), e);
+            restartContext.clearConnectionError();
+        } catch (ExecutionException | TimeoutException e) {
+            maybeTcpProbe(podId, e, restartContext);
         }
         int id = controllerNode == null || Node.noNode().equals(controllerNode) ? -1 : controllerNode.id();
         log.debug("controller is {}", id);
         return id;
     }
 
+    /**
+     * If we've already had trouble connecting to this broker try to probe whether the connection is
+     * open on the broker; if it's not then maybe throw a ForceableProblem to immediately force a restart.
+     * This is an optimization for brokers which don't seem to be running.
+     */
+    private void maybeTcpProbe(int podId, Exception executionException, RestartContext restartContext) throws ForceableProblem {
+        if (restartContext.connectionError() + numPods * 120_000L >= System.currentTimeMillis()) {
+            try {
+                log.debug("Probing TCP port due to previous problems connecting to pod {}", podId);
+                // do a tcp connect and close (with a short connect timeout)
+                tcpProbe(podName(podId), KafkaCluster.REPLICATION_PORT);
```

Comment:
If this `tcpProbe()` call does not throw an `IOException` then this method returns normally and we lose the fact that `executionException` was thrown and the caller will say the controller is `-1`. Is that correct, or should we rethrow `executionException` if no `IOException` was thrown?